src/Middleware.js

```diff
@@ -0,0 +1,104 @@
+import {MS, MOZ, WEBKIT, RULESET, KEYFRAMES, DECLARATION} from './Enum.js'
+import {match, charat, substr, strlen, sizeof, assign, replace, combine} from './Utility.js'
+import {tokenize} from './Tokenizer.js'
+import {serialize} from './Serializer.js'
+import {prefix} from './Prefixer.js'
+
+/**
+ * @param {function[]} collection
+ * @return {function}
+ */
+export function middleware (collection) {
+	var length = sizeof(collection)
+
+	return function (element, index, children, callback) {
+		var output = ''
+
+		for (var i = 0; i < length; i++)
+			output += collection[i](element, index, children, callback) || ''
+
+		return output
+	}
+}
+
+/**
+ * @param {function} callback
+ * @return {function}
+ */
+export function rulesheet (callback) {
+	return function (element) {
+		if (!element.root)
+			if (element = element.return)
+				callback(element)
+	}
+}
+
+/**
+ * @param {object} element
+ * @param {number} index
+ * @param {object[]} children
+ * @param {function} callback
+ */
+export function prefixer (element, index, children, callback) {
+	switch (element.type) {
+		case DECLARATION: element.return = prefix(element.value, element.length)
+			break
+		case KEYFRAMES:
+			return serialize([assign({}, element, {type: '', value: replace(element.value, '@', '@' + WEBKIT)})], callback)
```

Comment:
1. shouldnt those "copies" receive a meaningful `.type`? I would expect the "inserted" node to be resembling one that would be created when compiling
2. isn't usage of `assign` here creating a different hidden class than what is returned from `node` and thus making functions receiving this polymorphic which we should avoid?

Reply:
Point.2 is solved with the most recent commit, 

Point 1, that was intentional. The intention is to be able to distinguish between plugin generated rulesets and others, this signals that it is a no-op/completed node as far as a plugin is concerned so other plugins don't mistaken it for a source node and try to process it in their logic, it may make sense for other plugins to preserve the `.type` but in this is i thought it best not.